architecture.md

```diff
@@ -0,0 +1,174 @@
+
+## Information flow (in-depth)
+The current flow is as follows:
+```
+rustc -> librustc_save_analysis -> rls_data -> rls_analysis -> rls
+```
+
+### librustc_save_analysis
+
+The Rust compiler includes the `librustc_save_analysis` crate, which allows to dump the knowledge about the currently compiled crate. The main entry point is [`process_crate`](https://github.com/rust-lang/rust/blob/7164a9f151a56316a382d8bc2b15ccf373e129ca/src/librustc_save_analysis/lib.rs#L1119), which walks the post-macro-expansion AST and [saves](https://github.com/rust-lang/rust/blob/7164a9f151a56316a382d8bc2b15ccf373e129ca/src/librustc_save_analysis/lib.rs#L1146) the collected knowledge either by [dumping to a JSON file](https://github.com/rust-lang/rust/blob/7164a9f151a56316a382d8bc2b15ccf373e129ca/src/librustc_save_analysis/lib.rs#L1074-L1090) or by [calling back with resulting data structure](https://github.com/rust-lang/rust/blob/7164a9f151a56316a382d8bc2b15ccf373e129ca/src/librustc_save_analysis/lib.rs#L1092-L1117).
+
+### rls_data
+
+As mentioned previously, the returned data structure is [`rls_data::Analysis`](https://github.com/rust-dev-tools/rls-data/blob/9edbe8b4947c10ef670c4723be375c6944cab640/src/lib.rs#L30-L48) inside the [`rls_data`](https://github.com/rust-dev-tools/rls-data) crate:
+```rust
+/// Basically a type alias, we refer to nodes with HIR ids.
+/// All of the below nodes are either identified by or refer to those IDs.
+type Id = rustc::hir::def_id::DefId;
+
+pub struct Analysis {
+    ...
+    /// Contains rustc invocation producing this save-analysis data. Currently
+    /// used to support RLS in custom build system context, namely Buck).
+    pub compilation: Option<CompilationOptions>,
+    /// Path to current crate root file and current crate `GlobalCrateId`. Also
+    /// includes externally referred crates' `GlobalCrateId` and their local
+    /// `CrateNum` index as stored by rustc from this crate compilation PoV.
+    pub prelude: Option<CratePreludeData>,
+    /// Nodes of use tree forests, incl. relation, span, optional alias value
+    /// and kind (extern crate, simple use or glob).
+    pub imports: Vec<Import>,
+    /// Main data nodes. Roughly correspond to post-expansion AST nodes, incl.
+    /// span, qualified name, relations, optional signature (if a function),
+    /// docs and attributes.
+    pub defs: Vec<Def>,
+    /// Nodes for `impl` items, incl. kind (inherent, trait impl, ...), span,
+    /// children ids, docs, attributes and signature.
+    pub impls: Vec<Impl>,
+    /// Span which refers an `Id` of function/module/type/variable kind.
+    pub refs: Vec<Ref>,
+    /// Contains callsite and callee span along with macro qualified name.
+    pub macro_refs: Vec<MacroRef>,
+    /// Impl/SuperTrait relation between `Id`s with an associated span.
+    pub relations: Vec<Relation>,
+}
+```
+
+### rls_analysis
+
+This crate is responsible for loading and stitching multiple of
```

Comment:
and also this section:)
https://github.com/rust-dev-tools/rls-analysis

Reply:
ditto